File: swinehd_g1/src/main/java/playground/swine/DXTDecompress.java
package playground.swine;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.channels.FileChannel;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;

/*
 *	https://docs.microsoft.com/en-us/windows/desktop/direct3d10/d3d10-graphics-programming-guide-resources-block-compression
 */
public class DXTDecompress {

	public static void main(String[] args) throws IOException {
		Path root = new File("D:/Temp/190609_swine/remaster").toPath();

		// Find and convert all DXT files in the root folder
		Files.walk(root).forEach(path -> {
			File file = path.toFile();

			if (file.isDirectory() || !file.getName().endsWith(".dxt"))
				return;

			System.out.println(file);

			try {
				FileChannel input = new FileInputStream(file).getChannel();

				// Read file header
				ByteBuffer header = ByteBuffer.allocate(20);

				input.read(header);
				header.flip();

				// Verify stormregion signature
				int major = header.getInt();
				int minor = header.getInt();
				int type  = header.getInt();

				assert major == 0x53721A1B;
				assert minor == 0x0D0A870A;
				assert type  == 0x54455854;

				// Load texture data to memory
				header.order(ByteOrder.LITTLE_ENDIAN);

				int content = header.getInt();
				int unkn0   = header.getInt();

				ByteBuffer buffer = ByteBuffer.allocate(content).order(ByteOrder.LITTLE_ENDIAN);

				input.read(buffer);
				buffer.flip();

				int sizeX  = buffer.getInt();
				int sizeY  = buffer.getInt();
				int format = buffer.getInt();

				// Decode first mipmap level
				File into = new File(file.getParent(), file.getName().replace(".dxt", ".png"));

				ImageIO.write(decodeDXT(buffer, format, sizeX, sizeY), "PNG", into);

				// Done!
			} catch (IOException err) {
				throw new RuntimeException("Failed to decode file " + path, err);
			}
		});
	}

	public static final int DXT1 = 0x31545844;
	public static final int DXT5 = 0x35545844;

	static int unpack565(int color) {
		int R = (color >> 11) & 0x1F;
		int G = (color >> 5 ) & 0x3F;
		int B = (color >> 0 ) & 0x1F;

		// Upscale to 8 bit
		R = (R << 3) | (R >> 2);
		G = (G << 2) | (G >> 4);
		B = (B << 3) | (B >> 2);

		return (R << 16) | (G << 8) | B;
	}

	static int interpolate(int color0, int m0, int color1, int m1) {
		int total = m0 + m1;

		int A = ((color0 >> 24) & 0xFF) * m0 + ((color1 >> 24) & 0xFF) * m1;
		int R = ((color0 >> 16) & 0xFF) * m0 + ((color1 >> 16) & 0xFF) * m1;
		int G = ((color0 >> 8)  & 0xFF) * m0 + ((color1 >> 8)  & 0xFF) * m1;
		int B = ((color0 >> 0)  & 0xFF) * m0 + ((color1 >> 0)  & 0xFF) * m1;

		return (A / total) << 24 | (R / total) << 16 | (G / total) << 8 | (B / total);
	}

	// TODO: This makes it a tiny bit more efficient, but NOT thread safe/reentrant
	final static int[] alpha = new int[8];
	final static int[] color = new int[4];

	final static int[] texel = new int[16];

	static void decodeAlpha(ByteBuffer buffer, int[] argb) {

		// Build alpha palette
		alpha[0] = Byte.toUnsignedInt(buffer.get());
		alpha[1] = Byte.toUnsignedInt(buffer.get());

		// Interpolate alpha range
		if (alpha[0] > alpha[1]) {
			for (int idx = 2; idx < 8; idx++) {
				alpha[idx] = (alpha[0] * (8 - idx) + alpha[1] * (idx -1)) / 7;
			}
		} else {
			for (int i = 2; i < 6; i++) {
				alpha[i] = (alpha[0] *  + alpha[1] * (i -1)) / 5;
			}

			alpha[6] = 0;
			alpha[7] = 255;
		}

		// Unpack pixel alpha from indices
		for (int off = 0; off < 4*4; off += 8) {
			long packed = buffer.get() | (buffer.get() << 8) | (buffer.get() << 16);

			for (int pixel = 0; pixel < 8; pixel++) {
				int select = (int) (packed >> 3 * pixel) & 7;

				argb[off + pixel] = alpha[select] << 24;
			}
		}
	}

	static void decodeDiffuse(ByteBuffer colors, ByteBuffer indices, int[] argb, int alpha) {
		color[0] = (alpha << 24) | unpack565(colors.getShort());
		color[1] = (alpha << 24) | unpack565(colors.getShort());

		if (color[0] > color[1]) {
			color[2] = interpolate(color[0], 2, color[1], 1);
			color[3] = interpolate(color[0], 1, color[1], 2);
		} else {
			color[2] = interpolate(color[0], 1, color[1], 1);
			color[3] = 0x0;
		}

		// color[2] = interpolate(color[0], 1, color[1], 2);
		// color[3] = interpolate(color[0], 2, color[1], 1);

		for (int off = 0; off < 4*4; off += 4) {
			int packed = indices.get();

			argb[off + 0] |= color[(packed >> 0) & 3];
			argb[off + 1] |= color[(packed >> 2) & 3];
			argb[off + 2] |= color[(packed >> 4) & 3];
			argb[off + 3] |= color[(packed >> 6) & 3];
		}
	}

	private static void validateImage(int sizeX, int sizeY, int format) {
		if (sizeX % 4 != 0)
			throw new IllegalArgumentException("sizeX");

		if (sizeY % 4 != 0)
			throw new IllegalArgumentException("sizeY");

		switch (format) {
			case DXT1:
			case DXT5:
				break;

			default:
				throw new IllegalArgumentException("Unsupported image format " + Integer.toHexString(format));
		}
	}

	public static BufferedImage decodeDXT(ByteBuffer buffer, int format, int sizeX, int sizeY) {
		validateImage(sizeX, sizeY, format);

		BufferedImage image = new BufferedImage(sizeX, sizeY, BufferedImage.TYPE_INT_ARGB);

		// All reads expect little endian order
		buffer.order(ByteOrder.LITTLE_ENDIAN);

		for (int y = 0; y < sizeY; y += 4) {
			for (int x = 0; x < sizeX; x += 4) {
				Arrays.fill(texel, 0);

				if (format == DXT5) {
					decodeAlpha(buffer, texel);
					decodeDiffuse(buffer, buffer, texel, 0);
				} else {
					decodeDiffuse(buffer, buffer, texel, 255);
				}

				image.setRGB(x, y, 4, 4, texel, 0, 4);
			}
		}

		return image;
	}

	public static BufferedImage decodeBifurcatedDXT1(ByteBuffer buffer, int sizeX, int sizeY) {
		validateImage(sizeX, sizeY, DXT1);

		BufferedImage image = new BufferedImage(sizeX, sizeY, BufferedImage.TYPE_INT_ARGB);

		// This DXT buffer is bifurcated, color first, indices next
		int dataSize = (sizeX /4) * (sizeY /4) * 8;

		ByteBuffer color   = buffer.slice().order(ByteOrder.LITTLE_ENDIAN);
		ByteBuffer indices = buffer.slice().order(ByteOrder.LITTLE_ENDIAN);

		indices.position(dataSize /2);

		for (int y = 0; y < sizeY; y += 4) {
			for (int x = 0; x < sizeX; x += 4) {
				Arrays.fill(texel, 0xFF000000);

				decodeDiffuse(color, indices, texel, 255);

				image.setRGB(x, y, 4, 4, texel, 0, 4);
			}
		}

		// Simulate progress on the buffer
		buffer.position(buffer.position() + dataSize);

		return image;
	}

}

File: swinehd_g1/src/main/experiment/DXTDecompress.cpp
#include <fstream>
#include <iostream>
#include <vector>

constexpr int DXT1 = 0x31545844;
constexpr int DXT5 = 0x35545844;

int unpack565(int color) {
  int R = (color >> 11) & 0x1F;
  int G = (color >> 5) & 0x3F;
  int B = (color >> 0) & 0x1F;

  // Upscale to 8 bit
  R = (R << 3) | (R >> 2);
  G = (G << 2) | (G >> 4);
  B = (B << 3) | (B >> 2);

  return (R << 16) | (G << 8) | B;
}

int interpolate(int color0, int m0, int color1, int m1) {
  int total = m0 + m1;

  int A = ((color0 >> 24) & 0xFF) * m0 + ((color1 >> 24) & 0xFF) * m1;
  int R = ((color0 >> 16) & 0xFF) * m0 + ((color1 >> 16) & 0xFF) * m1;
  int G = ((color0 >> 8) & 0xFF) * m0 + ((color1 >> 8) & 0xFF) * m1;
  int B = ((color0 >> 0) & 0xFF) * m0 + ((color1 >> 0) & 0xFF) * m1;

  return (A / total << 24) | (R / total << 16) | (G / total << 8) | (B / total << 0);
}

int main(int argc, char** argv) {
  // Find and convert all DXT files in the root folder
  std::string root = "D:/Temp/190609_swine/remaster";
  std::vector<std::string> files = find_files_in_directory(root, ".dxt");
  for (const auto& file : files) {
    std::cout << file << std::endl;

    std::ifstream input(file, std::ios::binary);

    // Read file header
    std::vector<char> header(20);
    input.read(header.data(), header.size());

    // Verify stormregion signature
    int major = *reinterpret_cast<int*>(header.data());
    int minor = *reinterpret_cast<int*>(header.data() + 4);
    int type = *reinterpret_cast<int*>(header.data() + 8);

    assert(major == 0x53721A1B);
    assert(minor == 0x0D0A870A);
    assert(type == 0x54455854);

    // Load texture data to memory
    int content = *reinterpret_cast<int*>(header.data() + 12);
    int unkn0 = *reinterpret_cast<int*>(header.data() + 16);
    std::vector<char> buffer(content);
    input.read(buffer.data(), buffer.size());

    int sizeX = *reinterpret_cast<int*>(buffer.data());
    int sizeY = *reinterpret_cast<int*>(buffer.data() + 4);
    int format = *reinterpret_cast<int*>(buffer.data() + 8);

    // Decode first mipmap level
    std::string output_file = file.substr(0, file.size() - 4) + ".png";
    decodeDXT(buffer.data() + 12, format, sizeX, sizeY, output_file);

    // Done!
  }

  return 0;
}

void decodeDXT(const char* data, int format, int sizeX, int sizeY, const std::string& output_file) {
  if (format == DXT1) {
    // Create output image
    std::vector<int> pixels(sizeX * sizeY);

    int block_count_x = (sizeX + 3) / 4;
    int block_count_y = (sizeY + 3) / 4;

    // Decode blocks
    for (int y = 0; y < block_count_y; y++) {
      for (int x = 0; x < block_count_x; x++) {
        int block_index = y * block_count_x + x;
        int block_offset = block_index * 8;

        int color0 = *reinterpret_cast<const int*>(data + block_offset + 0);
        int color1 = *reinterpret_cast<const int*>(data + block_offset + 4);

        int codes = *reinterpret_cast<const int*>(data + block_offset + 8);

        int colors[4];
        colors[0] = unpack565(color0);
        colors[1] = unpack565(color1);

        if (color0 > color1) {
          colors[2] = interpolate(colors[0], 2, colors[1], 1);
          colors[3] = interpolate(colors[0], 1, colors[1], 2);
        } else {
          colors[2] = interpolate(colors[0], 1, colors[1], 1);
          colors[3] = 0;
        }

        for (int i = 0; i < 4; i++) {
          for (int j = 0; j < 4; j++) {
            int index = codes & 0x3;
            codes >>= 2;

            int pixel_x = x * 4 + j;
            int pixel_y = y * 4 + i;
            if (pixel_x >= sizeX || pixel_y >= sizeY)
              continue;

            pixels[pixel_y * sizeX + pixel_x] = colors[index];
          }
        }
      }
    }

    save_image(pixels, sizeX, sizeY, output_file);
  } else if (format == DXT5) {
    // Create output image
    std::vector<int> pixels(sizeX * sizeY);

    int block_count_x = (sizeX + 3) / 4;
    int block_count_y = (sizeY + 3) / 4;

    // Decode blocks
    for (int y = 0; y < block_count_y; y++) {
      for (int x = 0; x < block_count_x; x++) {
        int block_index = y * block_count_x + x;
        int block_offset = block_index * 16;

        int alpha0 = *reinterpret_cast<const char*>(data + block_offset + 0);
        int alpha1 = *reinterpret_cast<const char*>(data + block_offset + 1);

        int codes = *reinterpret_cast<const int*>(data + block_offset + 8);

        int alphas[8];
        alphas[0] = alpha0;
        alphas[1] = alpha1;
        if (alpha0 > alpha1) {
          for (int i = 0; i < 6; i++)
            alphas[i + 2] = (6 - i) * alpha0 / 6 + i * alpha1 / 6;
        } else {
          for (int i = 0; i < 4; i++)
            alphas[i + 2] = (4 - i) * alpha0 / 4 + i * alpha1 / 4;
          alphas[6] = 0;
          alphas[7] = 255;
        }

        int color0 = *reinterpret_cast<const int*>(data + block_offset + 4);
        int color1 = *reinterpret_cast<const int*>(data + block_offset + 12);

        int colors[4];
        colors[0] = unpack565(color0);
        colors[1] = unpack565(color1);

        if (color0 > color1) {
          colors[2] = interpolate(colors[0], 2, colors[1], 1);
          colors[3] = interpolate(colors[0], 1, colors[1], 2);
        } else {
          colors[2] = interpolate(colors[0], 1, colors[1], 1);
          colors[3] = 0;
        }

        for (int i = 0; i < 4; i++) {
          for (int j = 0; j < 4; j++) {
            int index = codes & 0x3;
            codes >>= 2;

            int pixel_x = x * 4 + j;
            int pixel_y = y * 4 + i;
            if (pixel_x >= sizeX || pixel_y >= sizeY)
              continue;

            int color = colors[index];
            int alpha_index = (codes >> (3 * (4 - i - 1))) & 0x7;
            int alpha = alphas[alpha_index];

            pixels[pixel_y * sizeX + pixel_x] = (alpha << 24) | (color & 0xFFFFFF);
          }
        }
      }
    }

    save_image(pixels, sizeX, sizeY, output_file);
  }
}

void save_image(const std::vector<int>& pixels, int width, int height, const std::string& file_name) {
  // ...
}

std::vector<std::string> find_files_in_directory(const std::string& directory, const std::string& extension) {
  // ...
}
